keps/sig-api-machinery/5051-server-side-apply-unsetting-fields/README.md

```diff
@@ -0,0 +1,966 @@
+  {name: "b", k8s_io__value: unset}
+]
+```
+
+Result value:
+
+```yaml
+field: [
+  {name: "a", value: 1}
+]
+```
+
+Result field management:
+
+```yaml
+fieldManager1: 
+  spec.field[name=a]
+
+fieldManager2: 
+  spec.field[name=b]
+```
+
+#### Apply (force=true) to unset an owned granular map value
+
+Existing value:
+
+```yaml
+field: {"a": 1, "b": 2}
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Force apply configuration:
+
+```yaml
+field: {"b": {k8s_io__value: unset}}
+```
+
+Result value:
+
+```yaml
+field: {"a": 1}
+```
+
+Result field management:
+
+```yaml
+fieldManager2:
+  field
+```
+
+#### Apply (force=false) to an owned field
+
+Existing value:
+
+```yaml
+spec:
+  field: 1
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Apply configuration:
+
+```yaml
+spec:
+  field: {k8s_io__value: unset}
+```
+
+Apply conflicts:
+
+```yaml
+field owned by fieldManager1
+```
+
+#### Apply (force=false) to a unset field that is owned
+
+Existing value:
+
+```yaml
+spec:
+  # empty
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Apply configuration:
+
+```yaml
+spec:
+  field: "xyz"
+```
+
+Apply conflicts:
+
+```yaml
+field owned by fieldManager1
+```
+
+#### Apply (force=false) to unset a field that is owned but already unset
+
+Existing value:
+
+```yaml
+spec:
+```
+
+Existing field management:
+
+```yaml
+fieldManager1: 
+  spec.field
+```
+
+Apply configuration:
+
+```yaml
+spec:
+  field: {k8s_io__value: unset}
+```
+
+Result value:
+
+```yaml
+spec:
+```
+
+Result field management:
+
+```yaml
+Shared ownership.
+
+fieldManager1: 
+  spec.field
+
+fieldManager2: 
+  spec.field
+```
+
+#### Apply to a field with a default value
+
+Existing value:
+
+```yaml
+spec:
+  field: defaultValue
+```
+
+Existing field management:
+
+```yaml
+fieldManager1:
+  # no owned fields
+```
+
+Force apply configuration:
+
+```yaml
+spec:
+  field: {k8s_io__value: unset}
+```
+
+Result value:
+
+```yaml
+spec:
+  field: defaultValue
+```
+
+Result field management:
+
+```yaml
+fieldManager2:
+  spec.field
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Defaulting considerations
+
+1. A field that is not explicitly set, but is defaulted, is unowned (existing behavior)
+1. If a field is unset using a marker, defaulting still applies
+   - Consequence: A field manager that owns a unset value ends up owning the defaulted value
+   - Consequence: If two field managers unset the same value, they share ownership of the field
+   - Caveat: Non-declarative defaulting, such as defaulting that is performed by the strategy, or
+    admission control, is not detectable by server side apply, and will result in conflicts between
+	field managers even for cases where the defaulting SHOULD result in shared field ownership.
+	This is a pre-existing problem between defaulting and server side apply but is more likely
+	to occur with this enhancement.
```

Comment:
Can you give a more concrete example? When you say "not detectable", what does that mean? Which kind of conflict? An update conflict deep in the storage stack?

Reply:
In general, I understand that admission defaulting behaves different than deserializer defaulting. What does that mean for your usecases above?

Reply:
Good question. The case I'm thinking of is:

- FieldManager 1: Applies field X to A
  - Apply succeeds, FieldManager 1 owns X
- Field X is set to B _after_ deserialization defaulting (typically the strategy or admission chain)
- FieldManager 2: Applies field X to A
- Apply conflict

This differs from deserialization defaulting, which will result in shared ownership of the field.

I'll find or open an issue explaining and link to it from this section of the KEP.

Reply:
Actually, I think this can only happen today with mutating admission, but could happen in the future as a result of defaulting as a result of this feature.  https://github.com/kubernetes/kubernetes/issues/129960 has details. I'l include a link in the KEP now.